File: c/p2379.c
#include <stdio.h>
#include <string.h>

int minimumRecolors(char* blocks, int k) {
    int length = strlen(blocks);
    int blackCount = 0;
    int highest = 0;

    for (int i = 0; i < k; i++) {
        if (blocks[i] == 'B') {
            highest += 1;
            blackCount += 1;
        }
    }

    printf("start : %d\n", highest);

    int current = highest;
    for (int i = k; i < length; i++) {
        if (blocks[i] == 'B') {
            blackCount += 1;
            current += 1;
        }

        if (blocks[i - k] == 'B') {
            current -= 1;
        }

        if (current > highest) {
            highest = current;
        }
        printf("%c : %c\n", blocks[i - k], blocks[i]);
    }

    printf("highest : %d\n", highest);

    int r = k - highest;

    if (r < 0) {
        return 0;
    }
    return r;
}

int main() {
    // printf("%d\n", minimumRecolors("WBBWWBBWBW", 7));
    // printf("%d\n", minimumRecolors("WBWBBBW", 2));
    printf("%d\n", minimumRecolors("WWBBBWBBBBBWWBWWWB", 16));
    printf("%d\n", minimumRecolors("ABCDEFGHIJKLMNOPQR", 16));
}

File: src/p2379.rs
pub fn run() {
    for i in [
        ("WBBWWBBWBW", 7),
        ("WBWBBBW", 2)
    ] {
        println!("{}", minimum_recolors(i.0.to_string(), i.1));
    }
}

pub fn minimum_recolors(blocks: String, k: i32) -> i32 {
    let k = k as usize;
    let chars = blocks.chars().collect::<Vec<char>>();
    let length = chars.len();
    let mut highest = 0i32;

    for i in 0..k {
        if chars[i] == 'B' {
            highest += 1;
        }
    }

    let mut current = highest;
    for i in k..length {
        if chars[i] == 'B' {
            current += 1;
        }

        if chars[i - k] == 'B' {
            current -= 1;
        }

        highest = highest.max(current);
    }
    
    let r = (k as i32) - highest;
    if r < 0 {
        return 0;
    }
    r
}
